File: python/day8.py
from pprint import pprint
from itertools import combinations

class Finder:

    def find_pair_nodes(self, r, c, dr, dc):
        d = 0
        while True:
            nr = r+(dr*d)
            nc = c+(dc*d)
            if not(nr in range(self.height) and nc in range(self.width)):
                break
            self.harmonic_nodes.add((nr,nc))
            if d == 1:
                self.nodes.add((nr,nc))
            d += 1

    def find_nodes(self):    
        for atype in self.ants:        
            for pair in set(combinations(self.ants[atype], 2)):                
                ((r1, c1),(r2,c2)) = pair
                dr = r1-r2
                dc = c1-c2

                self.find_pair_nodes(r1, c1, dr, dc)
                self.find_pair_nodes(r2, c2, -dr, -dc)
            
    def find_ants(self):
        ants = {}        
        for r,line in enumerate(input):        
            for c,a in enumerate(line):
                if a != '.':
                    if not a in ants:
                        ants[a]=[]
                    ants[a].append((r,c))
        self.ants = ants
        
        
    def run(self, input):        
        self.input = input        
        self.height = len(input)
        self.width = len(input[0])
        self.nodes = set()
        self.harmonic_nodes = set()
        
        self.find_ants()
        self.find_nodes()

        return(len(self.nodes),len(self.harmonic_nodes))

def day8(input):
    f = Finder()
    return f.run(input)

#14 34
input = open("src/bin/inputs/day8.test1.txt", "r").read().splitlines()
pprint(day8(input))

#276 991
input = open("src/bin/inputs/day8.txt", "r").read().splitlines()
pprint(day8(input))

File: src/bin/day8.rs
use itertools::Itertools;
use std::collections::{HashMap, HashSet};

const INPUT: &str = include_str!("inputs/day8.txt");


fn find_ants(map: &[&[u8]]) -> HashMap<u8, Vec<(usize,usize)>> {
    let mut ants = HashMap::<u8, Vec<_>>::new();
    for r in 0..map.len() {
        for c in 0..map[0].len() {
            if map[r][c] != b'.' {
                ants.entry(map[r][c]).or_default().push((r,c));
            }
        }
    }
    ants
}

fn find_pair_nodes( nodes: &mut HashSet<(usize,usize)>,
                    harmonic_nodes: &mut HashSet<(usize,usize)>,
                    r:usize, c:usize, dr:isize, dc:isize) {
    let mut d = 0;
    loop {
        let nr = r+(dr*d);
        let nc = c+(dc*d);
        if !((0..self.height).contains(&nr) && (0..).contains(&nc)) {
            break
        }
        harmonic_nodes.insert((nr,nc));
        if d == 1 {
            nodes.insert((nr,nc));      
        }
        d += 1;
    }
}

fn find_nodes(map: &[&[u8]], ants: &HashMap<u8, Vec<(usize,usize)>>) {
    let mut nodes: HashSet<(usize,usize)> = HashSet::new();
    let mut harmonic_nodes: HashSet<(usize,usize)> = HashSet::new();
    for (_,v) in ants.iter() {
        for pairs in v.iter().combinations(2) {
            let(r1,c1) = *pairs[0];
            let(r2,c2) = *pairs[1];
            let dr = (r1-r2) as isize;
            let dc = (c1-c2) as isize;

            //find_pair_nodes(&mut nodes, &mut harmonic_nodes, r1, c1, dr, dc);
            //find_pair_nodes(&mut nodes, &mut harmonic_nodes, r2, c2, 0-dr, 0-dc);
        }
    }
}


fn day8(input: &str) -> (usize, usize) {
    let map = input.lines().map(|x| x.as_bytes()).collect::<Vec<_>>();

    let ants = find_ants(&map);
    find_nodes(&map, &ants);

    (input.len(), input.len())
}

fn main() {
    let (p1, p2) = day8(INPUT);
    println!("{}\n{}", p1, p2);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test1() {
        let input = include_str!("inputs/day8.test1.txt");
        let (p1, p2) = day8(input);

        assert_eq!(p1, 14);
        assert_eq!(p2, 34);
    }

    #[test]
    fn test_main() {
        let (p1, p2) = day8(INPUT);

        assert_eq!(p1, 276);
        assert_eq!(p2, 991);
    }
}
